x-pack/metricbeat/module/containerd/cpu/metricset.go

```diff
@@ -0,0 +1,243 @@
+				e.RootFields.DeepUpdate(common.MapStr{
+					"container": containerFields,
+				})
+			} else {
+				e.RootFields = common.MapStr{
+					"container": containerFields,
+				}
+			}
+		}
+		if m.calcPct {
+			contCpus, ok := perContainerCpus[cID]
+			if !ok {
+				contCpus = 1
+			}
+			// calculate system usage delta
+			systemUsageDelta := float64(systemTotalNs) - m.preSystemCpuUsage
+
+			// Calculate cpu total usage percentage
+			cpuUsageTotal, err := event.GetValue("usage.total.ns")
+			if err == nil {
+				cpuUsageTotalPct := calcCpuTotalUsagePct(cpuUsageTotal.(float64), systemUsageDelta,
+					float64(contCpus), cID, m.preContainerCpuTotalUsage)
+				m.Logger().Debugf("cpuUsageTotalPct for %+v is %+v", cID, cpuUsageTotalPct)
+				e.MetricSetFields.Put("usage.total.pct", cpuUsageTotalPct)
+				// Update values
+				m.preContainerCpuTotalUsage[cID] = cpuUsageTotal.(float64)
+			}
+
+			// Calculate cpu kernel usage percentage
+			cpuUsageKernel, err := event.GetValue("usage.kernel.ns")
+			if err == nil {
+				cpuUsageKernelPct := calcCpuKernelUsagePct(cpuUsageKernel.(float64), systemUsageDelta,
+					float64(contCpus), cID, m.preContainerCpuKernelUsage)
+				m.Logger().Debugf("cpuUsageKernelPct for %+v is %+v", cID, cpuUsageKernelPct)
+				e.MetricSetFields.Put("usage.kernel.pct", cpuUsageKernelPct)
+				// Update values
+				m.preContainerCpuKernelUsage[cID] = cpuUsageKernel.(float64)
+			}
+
+			// Calculate cpu user usage percentage
+			cpuUsageUser, err := event.GetValue("usage.user.ns")
+			if err == nil {
+				cpuUsageUserPct := calcCpuUserUsagePct(cpuUsageUser.(float64), systemUsageDelta,
+					float64(contCpus), cID, m.preContainerCpuUserUsage)
+				m.Logger().Debugf("cpuUsageUserPct for %+v is %+v", cID, cpuUsageUserPct)
+				e.MetricSetFields.Put("usage.user.pct", cpuUsageUserPct)
+				// Update values
+				m.preContainerCpuUserUsage[cID] = cpuUsageUser.(float64)
+			}
+		}
+		if cpuId, err := event.GetValue("cpu"); err == nil {
+			perCpuNs, err := event.GetValue("usage.percpu.ns")
+			if err == nil {
+				key := fmt.Sprintf("usage.cpu.%s.ns", cpuId)
+				e.MetricSetFields.Put(key, perCpuNs)
+				e.MetricSetFields.Delete("cpu")
+				e.MetricSetFields.Delete("usage.percpu.ns")
+			}
+		}
+
+		if reported := reporter.Event(e); !reported {
+			return nil
+		}
+	}
+	m.preSystemCpuUsage = float64(systemTotalNs)
+	return nil
+}
+
+func setContCpus(event common.MapStr, perContainerCpus map[string]int) {
+	val, err := event.GetValue("id")
+	if err != nil {
+		return
+	}
+	cid := val.(string)
+	_, err = event.GetValue("usage.percpu.ns")
+	if err != nil {
+		return
+	}
+	if _, ok := perContainerCpus[cid]; ok {
+		perContainerCpus[cid] += 1
+	} else {
+		perContainerCpus[cid] = 1
+	}
+	//if percpu.(float64) != 0.0 {
+	//	if _, ok := perContainerCpus[cid]; ok {
+	//		perContainerCpus[cid] += 1
+	//	} else {
+	//		perContainerCpus[cid] = 1
+	//	}
+	//}
+}
+
+func calcCpuTotalUsagePct(cpuUsageTotal, systemUsageDelta, contCpus float64,
```

Comment:
I'm not clear on why these are three separate functions. They look really similar?